packages/grid/x-data-grid-pro/src/hooks/features/columnResize/useGridColumnResize.tsx

```diff
@@ -119,18 +134,144 @@ function getResizeDirection(element: HTMLElement, direction: Direction) {
+  const promise = React.useRef<ControllablePromise>();
+  const selector = () => gridVirtualizationColumnEnabledSelector(apiRef);
+  const value = useGridSelector(apiRef, selector);
+
+  React.useEffect(() => {
+    if (promise.current && value === false) {
+      promise.current.resolve();
+      promise.current = undefined;
+    }
+  });
+
+  const asyncCheck = () => {
+    if (!promise.current) {
+      if (selector() === false) {
+        return Promise.resolve();
+      }
+      promise.current = createControllablePromise();
+    }
+    return promise.current;
+  };
+
+  return asyncCheck;
+}
+
+/**
+ * Basic statistical outlier detection, checks if the value is `F * IQR` away from
+ * the Q1 and Q3 boundaries. IQR: interquartile range.
+ */
+function excludeOutliers(inputValues: number[], factor: number) {
+  if (inputValues.length < 4) {
+    return inputValues;
+  }
+
+  const values = inputValues.slice();
+  values.sort((a, b) => a - b);
+
+  const q1 = values[Math.floor(values.length * 0.25)];
+  const q3 = values[Math.floor(values.length * 0.75) - 1];
+  const iqr = q3 - q1;
+
+  // We make a small adjustment if `iqr < 5` for the cases where the IQR is
+  // very small (e.g. zero) due to very close by values in the input data.
+  // Otherwise, with an IQR of `0`, anything outside that would be considered
+  // an outlier, but it makes more sense visually to allow for this 5px variance
+  // rather than showing a cropped cell.
+  const deviation = iqr < 5 ? 5 : iqr * factor;
+
+  return values.filter((v) => v > q1 - deviation && v < q3 + deviation);
+}
+
+function extractColumnWidths(
+  apiRef: React.MutableRefObject<GridPrivateApiPro>,
+  options: AutosizeOptionsRequired,
+) {
+  type Result = Record<string, number>;
+
+  const headers = apiRef.current.columnHeadersContainerElementRef?.current!;
+  const container = apiRef.current.virtualScrollerRef?.current!;
+
+  const getHeader = (field: string) =>
+    headers.querySelector(`:scope > div > div > [data-field="${field}"][role="columnheader"]`);
+
+  const getCells = (field: string) =>
+    Array.from(
+      container.querySelectorAll(`:scope > div > div > div > [data-field="${field}"][role="cell"]`),
+    );
+
+  const widthByField = options.columns.reduce((result, column) => {
+    const cells = getCells(column.field);
+
+    const widths = cells.map((cell) => {
+      const style = window.getComputedStyle(cell, null);
+      const paddingWidth = parseInt(style.paddingLeft, 10) + parseInt(style.paddingRight, 10);
+      const contentWidth = (cell.firstElementChild?.scrollWidth ?? -1) + 1;
+      return paddingWidth + contentWidth;
+    });
+
+    const filteredWidths = options.excludeOutliers
+      ? excludeOutliers(widths, options.outliersFactor)
+      : widths;
+
+    if (options.includeHeaders) {
+      const header = getHeader(column.field);
+      if (header) {
+        const content = header.querySelector(`.${gridClasses.columnHeaderTitle}`)!;
```

Comment:
Oh, right. It won't work with the [`renderHeader`](https://deploy-preview-10333--material-ui-x.netlify.app/x/react-data-grid/column-header/#custom-header-renderer)
We need to measure children of `columnHeaderTitleContainerContent` element

Reply:
Fixed. We're now measuring the `columnHeaderTitleContainerContent`'s `scrollWidth`, which seems to return the correct value.

Reply:
I've added logic to use the title if it's there, because what we really want is the non-ellipsisized width of the children of `columnHeaderTitleContainerContent`, but I don't see a generic way to do that.